### ISSUE TITLE
Tuple unpacking from a type alias results in wrong union type

### ISSUE THREAD

**Issue by leddy231** [astral-sh/ty#1046]
### Summary

The following code snippet:
```python
type Thing = tuple[str, int]


def thing() -> Thing:
    return ("1", 2)


def foo():
    a, b = thing()
    a.lower()
```

Gives this error:
```python
warning[possibly-unbound-attribute]: Attribute `lower` on type `str | int` is possibly unbound
   |
 8 | def foo():
 9 |     a, b = thing()
10 |     a.lower()
   |     ^^^^^^^
   |
info: rule `possibly-unbound-attribute` is enabled by default
```
Because it thinks `a` is `str | int` when it should be just `str`. If the type alias is removed and the return type is explicitly `tuple[str, int]` it type checks correctly.

This seems like a regression in 0.0.1a18, it works correctly in 0.0.1a17

### Version

0.0.1a18

**Comment by AlexWaygood** [comment]
Thanks for the report! I can reproduce this by running `uvx ty check` on your snippet, but not on the ty playground. The ty playground is built from the `main` branch, so I suspect that after accidentally breaking this we might have accidentally fixed it again? Not sure...

**Comment by AlexWaygood** [comment]
`git bisect` indicates that the error went away again on `main` after https://github.com/astral-sh/ruff/commit/b892e4548e9b71949b1a6cb52d10ccb0c61e50d3

**Comment by AlexWaygood** [comment]
And `git bisect` indicates that the error was first _introduced_ in https://github.com/astral-sh/ruff/commit/13bdba5d28521e87373b72ce2e3f99a159ede644 -- so between those two commits, we had a bug.

It's much more obvious to me why https://github.com/astral-sh/ruff/commit/13bdba5d28521e87373b72ce2e3f99a159ede644 might have added a bug here (it looks like we could be missing a branch in `Type::try_iterate_with_mode()` for `Type::TypeAlias()`?) than why https://github.com/astral-sh/ruff/commit/b892e4548e9b71949b1a6cb52d10ccb0c61e50d3 would have fixed it again 😆

**Comment by carljm** [comment]
We unpack type aliases when they are added to a union or intersection, so I suspect that the latter commit causes us to pass the type through a union or intersection somewhere we didn't before.

It would still be good to add the case in `try_iterate_with_mode()`, I think; I'll see if I can make a test case that still reproduces.

**Comment by carljm** [comment]
Not too hard to repro if we avoid returning the alias from a function call: https://play.ty.dev/e66759b0-7d25-404c-b61d-baeae942faf3

Fix incoming.
